source/client-side-encryption/client-side-encryption.rst

```diff
@@ -591,6 +623,59 @@ the user. Refer:
+
+``EncryptedFieldConfigMap`` only applies to FLE 2.
+
+If a collection is present on both the ``EncryptedFieldConfigMap`` and ``schemaMap``, libmongocrypt_ will error on initialization. See :ref:`fle2-and-fle1-error`.
+
+If a collection is present on the ``EncryptedFieldConfigMap``, the behavior of ``CreateCollection()`` and ``Collection.Drop()`` is altered. See :ref:`fle2-createcollection-drop`.
+
+Automatic encryption in FLE 2 is configured with the ``EncryptedFieldConfig``.
+
+If a collection is not present on the ``EncryptedFieldConfig`` a server-side collection ``EncryptedFieldConfig`` may be used by libmongocrypt_.
+Drivers MUST include the following in the documentation for MongoClient:
+
+   Supplying an ``EncryptedFieldConfigMap`` provides more security than relying on an ``EncryptedFieldConfig`` obtained from the server.
+   It protects against a malicious server advertising a false ``EncryptedFieldConfig``.
+
+.. _fle2-createcollection-drop:
+
+FLE 2 ``CreateCollection()`` and ``Collection.Drop()``
+------------------------------------------------------
+
+A collection supporting FLE 2 requires an index and three additional collections.
+
+A call to a driver helper ``CreateCollection(collectionName, collectionOptions)`` must check if the collection namespace (``<databaseName>.<collectionName>``) is in ``AutoEncryptionOpts.EncryptedFieldConfigMap``.
+If it is, then do the following operations. If any of the following operations error, the remaining operations are not attempted:
+:
+
+- Create the collection with name ``EncryptedFieldConfig["escCollection"]`` using default options.
+  If ``EncryptedFieldConfig["escCollection"]`` is not set, use the collection name ``enxcol_.<collectionName>.esc``.
+  Creating this collection MUST NOT check if the collection namespace is in the ``AutoEncryptionOpts.EncryptedFieldConfigMap``.
+- Create the collection with name ``EncryptedFieldConfig["eccCollection"]`` using default options.
+  If ``EncryptedFieldConfig["eccCollection"]`` is not set, use the collection name ``enxcol_.<collectionName>.ecc``.
+  Creating this collection MUST NOT check if the collection namespace is in the ``AutoEncryptionOpts.EncryptedFieldConfigMap``.
+- Create the collection with name ``EncryptedFieldConfig["ecocCollection"]`` using default options.
+  If ``EncryptedFieldConfig["ecocCollection"]`` is not set, use the collection name ``enxcol_.<collectionName>.ecoc``.
+  Creating this collection MUST NOT check if the collection namespace is in the ``AutoEncryptionOpts.EncryptedFieldConfigMap``.
+- Create the collection ``collectionName`` with ``collectionOptions`` and the option ``encryptedFields`` set to the ``EncryptedFieldConfig``.
+- Create the the index ``{"__safeContent__": 1}`` on collection ``collectionName``.
+
+
+A call to a driver helper ``Collection.Drop()`` must check if the collection namespace (``<databaseName>.<collectionName>``) is in ``AutoEncryptionOpts.EncryptedFieldConfigMap``.
```

Comment:
Did we consider whether `drop` should use an ``EncryptedFieldConfig`` obtained from the server, as seems to be done for auto-encryption, according to the note:

> Supplying an ``EncryptedFieldConfigMap`` provides more security than relying on an ``EncryptedFieldConfig`` obtained from the server.

?

Reply:
Good question. I had not considered it.

Here is a proposal to use `listCollections` in Go:

```go
func (coll *Collection) Drop(ctx context.Context) error {
	efcMap := coll.db.client.encryptedFieldConfigMap
	if efcMap != nil {
		// Use listCollections to check the server for an EncryptedFieldConfig.
		efc, err := coll.getEncryptedFieldConfigFromServer(ctx)
		if err != nil {
			return err
		}
		if efc != nil {
			// Drop collection and state collections from EncryptedFieldConfig.
			return coll.dropEncryptedCollection(ctx, efc)
		}
	}
	// Do a normal Drop.
	return coll.dropHelper(ctx)
}
```

The behavior is gated behind an `EncryptedFieldConfigMap` being set. This assumes it is not desirable to run the `listCollections` in the Drop helper unless there is reason to believe the user has FLE 2 collections.

I am in favor of adding the call to `listCollections`. It does not add much complexity. Does this seem OK?

Reply:
Added a check in `Drop`. If `EncryptedFieldConfigMap` is set and the collection has no entry in the map, `listCollections` is run to check for an `EncryptedFieldConfig` on the server.